### PR TITLE
PowerSimulation v0.6

### LINKED ISSUE #154: add new network models from PowerModels

### LINKED ISSUE #195: review UC formulations
Review the UC formulations with respect to 

https://github.com/power-grid-lib/pglib-uc/blob/master/MODEL.pdf

### LINKED ISSUE #444: OperationsResults timestamps not updated when initial time is defined
The timestamps in OperationalProblemsResults don't match the forecast used when the initial_condition kwarg is used.

### LINKED ISSUE #446: Performance issue in axis_array_to_dataframe
The filter action in the line below causes a performance issue when running a simulation.  In a test with 292 columns it took about 16ms per iteration and so the function took 4.7 seconds.

```
function axis_array_to_dataframe(input_array::JuMP.Containers.SparseAxisArray)
    column_names = unique([(k[1], k[3]) for k in keys(input_array.data)])
    array_values = Vector{Vector{Float64}}(undef, length(column_names))
    for (ix, col) in enumerate(column_names)
        res = values(filter(v -> first(v)[[1, 3]] == col, input_array.data))
        array_values[ix] = PSI._jump_value.(res)
    end
    df = DataFrames.DataFrame(array_values, Symbol.(column_names))
    return df
end
```

### LINKED ISSUE #449: Recorder skipping initial condition update
Seems that the recorder isn't picking up the Initial Conditions updates. 

```
IS.@record :simulation InitialConditionUpdateEvent(
                Dates.DateTime(1970, 1, 1),
                key,
                ic,
                val,
                0,
            )
```

### LINKED ISSUE #450: Simulation initial conditions skip using ParameterJump
In the refactor of initial conditions the references to create ParameterRef's were lost. Currently, all initial conditions are kept from the beginning of the simulation.

### LINKED ISSUE #485: Certain functions should have an exclamation mark
I believe functions such as `device_semicontinuousrange` and `device_range` should probably be called `device_semicontinuousrange!` and `device_range!`, since they modify their first argument (`psi_container`) and don't return anything. The code is sometimes a little confusing because it looks like there's something missing.
If you agree with this I'll be happy to open a PR.

### PR DESCRIPTION
**Pull request by jd-lara**
closes #485 
closes #450 
closes #449 
closes #446
closes #444 
closes #195 
closes #154

### CODE REVIEW

File: src/devices_models/devices/renewable_generation.jl [hunk 1]
```diff
@@ -84,151 +86,68 @@ function reactivepower_constraints!(
...
-            variable_name(ACTIVE_POWER, R),
+    use_parameters::Bool,
+    use_forecasts::Bool,
+)
+    if (!use_parameters && !use_forecasts)
+        return DeviceRangeConstraintInputs(;
+            range_constraint_inputs = [RangeConstraintInputs(;
+                constraint_name = ACTIVE_RANGE,
+                variable_name = ACTIVE_POWER,
+                limits_func = x -> (min = 0.0, max = PSY.get_activepower(x)),
```
Comment (claytonpbarrows) [inline src/devices_models/devices/renewable_generation.jl:102]: I think this is wrong

Comment (jd-lara) [reply src/devices_models/devices/renewable_generation.jl:102]: This is ok, when no parameters and forecasts are used. The Constraint is based on the current operating point.

**Comment by jd-lara** [comment]
@daniel-thom and @claytonpbarrows clearly there is too much going on here. But a quick review overall will be good.

File: src/devices_models/devices/common/get_time_series.jl [hunk 2]
```diff
@@ -7,8 +7,11 @@ function get_time_series(
     @debug initial_time
     use_forecast_data = model_uses_forecasts(psi_container)
     time_steps = model_time_steps(psi_container)
-    @debug "device $(PSY.get_name(device)) $(PSY.has_forecasts(device)) forecast"
-    if use_forecast_data & PSY.has_forecasts(device)
+    has_forecasts = PSY.has_forecasts(device)
+    if !has_forecasts
+        @warn("device $(typeof(device)) $(PSY.get_name(device)) has forecasts = $(PSY.has_forecasts(device))")
+    end
```
Comment (daniel-thom) [inline src/devices_models/devices/common/get_time_series.jl:13]: Perhaps this shouldn't be logged if the user has not requested forecasts to be used.
Also, here is a suggestion to simplify the log message.

```suggestion
    if use_forecast_data && !has_forecasts
        @warn "$(summary(device)) does not have forecasts)"
    end
```

File: src/devices_models/devices/common/nodal_expression.jl [hunk 3]
```diff
@@ -0,0 +1,89 @@
+struct NodalExpressionSpec
+    forecast_label::String
+    # TODO: Remove this hack
```
Comment (daniel-thom) [inline src/devices_models/devices/common/nodal_expression.jl:3]: Should we submit an issue to track this?

File: src/network_models/network_constructor.jl [hunk 4]
```diff
@@ -6,13 +6,28 @@ function construct_network!(
     buses = PSY.get_components(PSY.Bus, sys)
     bus_count = length(buses)
 
-    get_slack_variables(psi_container.settings) &&
+    get_balance_slack_variables(psi_container.settings) &&
```
Comment (daniel-thom) [inline src/network_models/network_constructor.jl:9]: Minor:  Can we write these as if statements?  Being able to skip the `end` statement doesn't justify this, IMHO.

File: src/network_models/network_constructor.jl [hunk 5]
```diff
@@ -6,13 +6,28 @@ function construct_network!(
...
 
+function construct_network!(
+    psi_container::PSIContainer,
+    sys::PSY.System,
+    ::Type{AreaBalancePowerModel},
+)
+    area_mapping = PSY.get_aggregation_topology_mapping(PSY.Area, sys)
+    branches = get_available_components(PSY.Branch, sys)
+    if get_balance_slack_variables(psi_container.settings)
+        @warn("Slack Variables are not compatible with AreaBalancePowerModel")
```
Comment (daniel-thom) [inline src/network_models/network_constructor.jl:24]: Maybe this has already been covered; please delete if that is the case.  Should this call `error()` instead?

File: src/printing.jl [hunk 6]
```diff
@@ -79,18 +79,52 @@ function Base.show(io::IO, results::OperationsProblemResults)
...
     println(io, "\nResults")
-    println(io, "===============\n")
-
-    for (k, v) in results.variable_values
-        time = DataFrames.DataFrame(Time = results.time_stamp[!, :Range])
-        if size(time, 1) == size(v, 1)
-            var = hcat(time, v)
-        else
-            var = v
+    println(io, "========\n")
```
Comment (daniel-thom) [inline src/printing.jl:82]: Should this function be moved to IS?  As of now, no one would be able to implement this method there without causing a method overwrite.

File: src/printing.jl [hunk 7]
```diff
@@ -104,18 +138,50 @@ function Base.show(io::IO, ::MIME"text/plain", results::IS.Results)
...
-
 function Base.show(io::IO, ::MIME"text/html", results::IS.Results)
     println(io, "<h1>Results</h1>")
-    for (k, v) in results.variable_values
-        time = DataFrames.DataFrame(Time = results.time_stamp[!, :Range])
-        if size(time, 1) == size(v, 1)
-            var = hcat(time, v)
-        else
-            var = v
+    println(io, "<h2>Variables</h2>")
```
Comment (daniel-thom) [inline src/printing.jl:143]: ditto

File: src/services_models/agc.jl [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+    R_dn_emergency =
+        get_variable(psi_container, make_variable_name("area_emergency_reserve_dn"))
+
+    container = JuMPConstraintArray(undef, time_steps)
+    assign_constraint!(psi_container, "freque_response", container)
+
+    #container[1] = JuMP.@constraint(
+    #        psi_container.JuMPmodel,
+    #        frequency[1] == -inv_frequency_reponse * (sum(area_balance.data[:, 1]))
+    #    )
```
Comment (daniel-thom) [inline src/services_models/agc.jl:160]: Does something need to be done here?

File: src/PowerSimulations.jl [hunk 9]
```diff
@@ -122,20 +136,78 @@ export get_duals
...
 export get_var_index
 export get_result_variable
+export get_variable_names
 export configure_logging
 export show_recorder_events
 export list_simulation_events
 export show_simulation_events
 
+# Variables / Parameters
+export ACTIVE_POWER
```
[the remark this answers is not in the document]
Comment (jd-lara) [reply src/PowerSimulations.jl:155]: I am ok moving them. Aqua makes those tests for outdated exports.

File: src/devices_models/devices/common/add_parameters.jl [hunk 10]
```diff
@@ -1,29 +1,50 @@
...
     psi_container::PSIContainer,
-    ts_data::Vector{DeviceTimeSeries},
+    constraint_infos::Vector{DeviceTimeSeriesConstraintInfo},
     param_reference::UpdateRef,
     expression_name::Symbol,
     multiplier::Float64 = 1.0,
 )
     @assert model_has_parameters(psi_container)
     time_steps = model_time_steps(psi_container)
     ## TODO, replace multiplier
```
[the remark this answers is not in the document]
Comment (jd-lara) [reply src/devices_models/devices/common/add_parameters.jl:11]: Not relevant, will remove

File: src/devices_models/devices/common/nodal_expression.jl (hunk 3 above)
Comment (jd-lara) [reply src/devices_models/devices/common/nodal_expression.jl:3]: This hack is part of the Simulation refactor. I will clarify

File: src/printing.jl (hunk 6 above)
Comment (jd-lara) [reply src/printing.jl:82]: I don't see why we would send this to IS. It seems unnecessary and it is something it should be easy to adjust as the wording changes

File: src/services_models/agc.jl (hunk 8 above)
Comment (jd-lara) [reply src/services_models/agc.jl]: Stale code from development

File: src/devices_models/devices/renewable_generation.jl [hunk 11]
```diff
@@ -4,231 +4,162 @@ struct RenewableFullDispatch <: AbstractRenewableDispatchFormulation end
...
+    if !use_parameters && !use_forecasts
+        return DeviceRangeConstraintSpec(;
+            range_constraint_spec = RangeConstraintSpec(;
+                constraint_name = make_constraint_name(
+                    RangeConstraint,
+                    ActivePowerVariable,
+                    T,
+                ),
+                variable_name = make_variable_name(ActivePowerVariable, T),
+                limits_func = x -> (min = 0.0, max = PSY.get_active_power(x)),
```
Comment (claytonpbarrows) [inline src/devices_models/devices/renewable_generation.jl:113]: should this be `get_max_active_power(x)`?

**Comment by jd-lara** [comment]
@daniel-thom the updates to make consistent the accessing and variable naming are quite substantial. These can be refactored in a later PR since all the logic now works.

File: src/devices_models/devices/renewable_generation.jl (hunk 11 above)
Comment (jd-lara) [reply src/devices_models/devices/renewable_generation.jl:113]: No, this is the case for no_forecasts. It loads the current operating point in the system.

File: src/network_models/network_constructor.jl [hunk 12]
```diff
@@ -21,7 +21,7 @@ function construct_network!(
     area_mapping = PSY.get_aggregation_topology_mapping(PSY.Area, sys)
     branches = get_available_components(PSY.Branch, sys)
     if get_balance_slack_variables(psi_container.settings)
-        @warn("Slack Variables are not compatible with AreaBalancePowerModel")
+        throw(IS.ConflictingInputsError(""))
```
Comment (daniel-thom) [inline src/network_models/network_constructor.jl:24]: Needs a message

File: src/devices_models/devices/common/get_time_series.jl [hunk 13]
```diff
@@ -9,7 +9,7 @@ function get_time_series(
     time_steps = model_time_steps(psi_container)
     has_forecasts = PSY.has_forecasts(device)
     if !has_forecasts
-        @warn("device $(typeof(device)) $(PSY.get_name(device)) has forecasts = $(PSY.has_forecasts(device))")
+        @warn "$(summary(device)) does not have forecasts)"
```
Comment (daniel-thom) [inline src/devices_models/devices/common/get_time_series.jl:12]: Per my earlier comment, is it appropriate to log this warning when `use_forecast_data` is false?

**Pull request merged**